def help_file() {
    log.info """
    #######################################################################################
    ######################### DOWNLOAD FILES FROM BPA DATA PORTAL #########################
    #######################################################################################

        --yaml <PATH/TO/YAML/FILE>
                Path to the .yaml file generated by the data mapper query

        --sample_id SAMPLE_ID
                BPA organism grouping key of species for which to download data. 
                Input .jsonl or samplesheet will be filtered on this field so only files
                corresponding to the grouping key will be downloaded. 
                SAMPLE_ID should be of the form 'taxid12345', where 12345 is the NCBI 
                taxonomy id of the species

        --use_samplesheet
                Pull data from a samplesheet rather than a data mapper output file?
                Default is 'false'

        --jsonl <PATH/TO/JSONL/FILE>
                Path to the .jsonl(.gz) file outputted by the data mapper

        --samplesheet <PATH/TO/SAMPLESHEET>
                Path to a samplesheet with information on samples to be
                downloaded. This is a .csv with the following column headers:
                
                    sample_id       Sample ID on which data in the samplesheet will be
                                        filtered. Should correspond to an individual.

                    data_type       The type of data within each file. Currently
                                        supported types: 'PACBIO', 'Hi-C', 'ONT'

                    file_name       The name of the file to be downloaded

                    file_checksum   The md5sum of the file to be downloaded 

                    url             The URL from which to download the file

                    busco_lineage   The busco lineage to be used with the sample in
                                        downstream analyses

                    scientific_name Binomial name of the sample specimen
                    
                See assets/samplesheet_example.csv for an example

        --bpa_api_token BPA_API_TOKEN
                API token for BioPlatforms Australia, enables downloading of datasets
                that may not be publically accessible. Token can be created by logging 
                into https://data.bioplatforms.com/, then going to
                https://data.bioplatforms.com/user/<username>/api-tokens, and clicking 
                'Create API Token'. 

        --outdir <PATH/TO/OUTPUT/DIRECTORY>
                File path to where results should be stored
                Default is './results'

        --pacbio_data
                Does the sample_id have PacBio HiFi data files to download, or not?
                Default is 'false'

        --hic_data
                Does the sample_id have HiC data files to download, or not?
                Default is 'false'

        --ont_data
                Does the sample_id have Oxford Nanopore data files to download, or not?
                Default is 'false'

    #######################################################################################
    """.stripIndent()
}


// print help file if requested
if ( params.remove('help') ) {
    help_file()
    exit 0
}

// check no unexpected parameters were specified
allowed_params = [
    // pipeline inputs
    "yaml",
    "sample_id",
    "use_samplesheet",
    "jsonl",
    "samplesheet",
    "bpa_api_token",
    "outdir",
    "pacbio_data",
    "hic_data",
    "ont_data",

    // Pawsey options
    "max_cpus",
    "max_memory"
]

params.each { entry ->
    if ( !allowed_params.contains(entry.key) ) {
        println("The parameter <${entry.key}> is not known");
        exit 0;
    }
}

// check required parameters are specified; throw error if not
if ( !params.yaml ) { error(
    """
    ERROR: No yaml file provided: \'--sample_id\'
    """
)}

// if ( !params.sample_id ) { error(
//     """
//     ERROR: No organism grouping key provided: \'--sample_id\'
//     """
// )}

// if ( !params.use_samplesheet && !params.jsonl ) { error(
//     """
//     ERROR: No data mapper output file provided: \'--jsonl\'
//     """
// )}

// if ( !params.use_samplesheet && !file(params.jsonl).exists() ) { error(
//     """
//     ERROR: Data mapper output file provided by \'--jsonl\' does not exist
//     """
// )}

// if ( params.use_samplesheet && !params.samplesheet ) { error(
//     """
//     ERROR: Samplesheet mode selected (\'--use_samplesheet\'), but 
//     no samplesheet file provided: \'--samplesheet\'
//     """
// )}

// if ( params.use_samplesheet && !file(params.samplesheet).exists() ) { error(
//     """
//     ERROR: Samplesheet file provided by \'--samplesheet\' does not exist
//     """
// )}

// some files can be downloaded without a token, but better to have one
// amend in future if necessary
if ( !params.bpa_api_token ) { error(
    """
    ERROR: No BPA API token provided: \'--bpa_api_token\'
    """
)}

// if no data types are specified, throw error
if ( !params.pacbio_data && !params.hic_data && !params.ont_data ) { error(
    """
    ERROR: \'--pacbio_data\', \'--hic_data\', and \'--ont_data\' flags are all turned off.
    No data files will be downloaded. Please set at least one data flag.
    """
)}

// pull in scripts for relevant processes
include { JSON_TO_TSV } from './modules/json_to_tsv.nf'
include { DOWNLOAD_FILE as DOWNLOAD_FILE_PACBIO } from './modules/download_file.nf'
include { DOWNLOAD_FILE as DOWNLOAD_FILE_HIC } from './modules/download_file.nf'
include { DOWNLOAD_FILE as DOWNLOAD_FILE_ONT } from './modules/download_file.nf'


// actually run the workflow
workflow {

    READ_YAML(yaml)

    READ_YAML.out.input.view()

    READ_YAML.out.sd.view()
    // ################################
    // ### getting lists of samples ###
    // ################################


    // if ( !params.use_samplesheet ) {
    //     // set up channel for input jsonl file
    //     json_to_tsv_ch = Channel.fromPath(params.jsonl)

    //     // parse it to tsv format for legibility
    //     JSON_TO_TSV(json_to_tsv_ch)

    //     // read in all the rows of the new tsv file
    //     all_samples = JSON_TO_TSV.out.tsv
    //         .splitCsv(header:true, sep:'\t')
    //         //.view()

    // } else {
    //     all_samples = Channel.fromPath( params.samplesheet )
    //         .splitCsv(header:true)
    // }
    
    // // ##################################################
    // // ### get pacbio reads for sample_id of interest ###
    // // ##################################################

    // if ( params.pacbio_data ) {

    //     if ( !params.use_samplesheet ) {

    //         pacbio_samples = all_samples
    //             .filter { sample -> sample.organism_grouping_key == "${params.sample_id}" } // keeps only the samples for the species we want
    //             .filter { sample -> sample.platform == "PACBIO_SMRT" } // keeps only the PacBio samples
    //             .filter { sample -> sample.library_strategy == "WGS" } // keeps only the WGS ones - filters out longread RNA-seq 
    //                     // NOTE: also filters out a few samples with WGA library strategy
    //                     // if we want to keep these, maybe instead filter on "library_source: GENOMIC"
    //             .filter { sample -> sample.optional_file == "false" } // filters out any .subreads.bam files
    //             .map {sample -> [sample.organism_grouping_key, sample.file_name, sample.url, sample.file_checksum] }

    //     } else {
    //         pacbio_samples = all_samples
    //             .filter { sample -> sample.sample_id == "${params.sample_id}" }
    //             .filter { sample -> sample.data_type == "PACBIO" }
    //             .map {sample -> [sample.sample_id, sample.file_name, sample.url, sample.file_checksum] }
    //     }

    //     // if no PacBio Samples are found, throw an error and exit the process
    //     pacbio_samples.ifEmpty { error(
    //         """
    //         \'--pacbio_data\' is set as true, but no PacBio samples corresponding to sample id 
    //         \"${params.sample_id}\" could be found.
    //         Check sample information or turn off \'--pacbio_data\' flag
    //         """
    //         ) }
    //     //pacbio_samples.view()

    //     DOWNLOAD_FILE_PACBIO(pacbio_samples, 'hifi')

    // }


    // // ###############################################
    // // ### get hic reads for sample_id of interest ###
    // // ###############################################

    // if ( params.hic_data ) {

    //     if ( !params.use_samplesheet ) {

    //         hic_samples = all_samples
    //             .filter { sample -> sample.organism_grouping_key == "${params.sample_id}" }
    //             .filter { sample -> sample.library_strategy == "Hi-C" }
    //             .map { sample -> [sample.organism_grouping_key, sample.file_name, sample.url, sample.file_checksum] }

    //     } else {

    //         hic_samples = all_samples
    //             .filter { sample -> sample.sample_id == "${params.sample_id}" }
    //             .filter { sample -> sample.data_type == "Hi-C" }
    //             .map {sample -> [sample.sample_id, sample.file_name, sample.url, sample.file_checksum] }
                
    //     }

    //     hic_samples.ifEmpty { error(
    //         """
    //         \'--hic_data\' is flagged, but no Hi-C samples corresponding to sample id 
    //         \"${params.sample_id}\" could be found.
    //         Check sample information or turn off \'--hic_data\' flag
    //         """) }

    //     DOWNLOAD_FILE_HIC(hic_samples, 'hic')
    // }


    // // ###############################################
    // // ### get ont reads for sample_id of interest ###
    // // ###############################################

    // // NOTE: there is currently no nanopore data in the data mapper output

    // if ( params.ont_data ) {
    
    //     if ( !params.use_samplesheet ) {
    //         // may also need to filter on archive_type = fastq_pass to avoid getting failed reads or pod5 output
    //         // keep an eye on data mapper output to see what happens here
    //         ont_samples = all_samples
    //             .filter { sample -> sample.organism_grouping_key == "${params.sample_id}" }
    //             .filter { sample -> sample.library_strategy == "ONT" }
    //             .map { sample -> [sample.organism_grouping_key, sample.file_name, sample.url, sample.file_checksum] }

    //     } else {

    //             ont_samples = all_samples
    //             .filter { sample -> sample.sample_id == "${params.sample_id}" }
    //             .filter { sample -> sample.data_type == "ONT" }
    //             .map {sample -> [sample.sample_id, sample.file_name, sample.url, sample.file_checksum] }

    //     }

    //     ont_samples.ifEmpty { error(
    //         """
    //         \'--ont_data\' is flagged, but no ONT samples corresponding to sample id 
    //         \"${params.sample_id}\" could be found.
    //         Check sample information or turn off \'--ont_data\' flag
    //         """) }\

    //     DOWNLOAD_FILE_ONT(ont_samples, 'ont')
    // }
}